### PR TITLE
feat: support custom status render in qrcode

### PR DESCRIPTION
**Pull request by kiner-tang**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

#### Motivation

Currently, our QRCode has several states, but the display content for each state is immutable. However, in many cases, we need to customize the content for each state according to the design. Therefore, a `statusRender` is needed to customize the rendering for each state.

#### API

```typescript
export type QRStatus = 'active' | 'expired' | 'loading' | 'scanned';

type StatusRenderInfo = {
  status: Exclude<QRStatus, 'active'>;
  locale: Locale['QRCode'];
  onRefresh?: () => void;
};

{
    statusRender?: (oriNode: ReactNode, info: StatusRenderInfo) => ReactNode;
}
```

#### Preview

image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | support custom status render in qrcode |
| 🇨🇳 Chinese | 在二维码中支持自定义状态渲染器 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

### CODE REVIEW

File: components/qr-code/index.zh-CN.md [hunk 1]
```diff
@@ -48,8 +49,18 @@ tag: 5.1.0
...
 | status | 二维码状态 | `active \| expired \| loading \| scanned` | `active` | scanned: 5.13.0 |
+| statusRender | 自定义状态渲染器 | (oriNode: React.ReactNode, info: [StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode | - | 5.19.0 |
 | onRefresh | 点击"点击刷新"的回调 | `() => void` | - |
 
+### StatusRenderInfo
+
+```typescript
+type StatusRenderInfo = {
+  locale: Locale['QRCode'];
+  onRefresh?: () => void;
```
Comment (afc163) [inline components/qr-code/index.zh-CN.md:60]: 需要 status 。

File: components/qr-code/interface.ts [hunk 2]
```diff
@@ -19,4 +27,9 @@ export interface QRCodeProps extends QRProps, React.HTMLAttributes<HTMLDivElemen
   errorLevel?: 'L' | 'M' | 'Q' | 'H';
   status?: 'active' | 'expired' | 'loading' | 'scanned';
   onRefresh?: () => void;
+  statusRender?: {
+    expired?: (defaultNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+    loading?: (defaultNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+    scanned?: (defaultNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+  };
```
Comment (afc163) [inline components/qr-code/interface.ts:34]: 这个和 API 表格里的描述貌似不一样。

Comment (kiner-tang) [reply components/qr-code/interface.ts:34]: done

File: components/qr-code/index.zh-CN.md (hunk 1 above)
Comment (kiner-tang) [reply components/qr-code/index.zh-CN.md:61]: done

File: components/qr-code/interface.ts [hunk 3]
```diff
@@ -17,6 +28,11 @@ export interface QRCodeProps extends QRProps, React.HTMLAttributes<HTMLDivElemen
...
   bordered?: boolean;
   errorLevel?: 'L' | 'M' | 'Q' | 'H';
-  status?: 'active' | 'expired' | 'loading' | 'scanned';
+  status?: QRStatus;
   onRefresh?: () => void;
+  statusRender?: {
+    expired?: (oriNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+    loading?: (oriNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+    scanned?: (oriNode: ReactNode, info: StatusRenderInfo) => ReactNode;
+  };
```
Comment (afc163) [inline components/qr-code/interface.ts:37]: 我看还是不一样，表格里是一个 function，这里是一个 object

Comment (kiner-tang) [reply components/qr-code/interface.ts:37]: 是的，我抽空改一下文档

**Comment by afc163** [comment]
我意思是一个 function 更简单，三个就复杂了。

**Comment by kiner-tang** [comment]
> 我意思是一个 function 更简单，三个就复杂了。

理解错你的意思了，我调整一下

**Comment by kiner-tang** [comment]
为啥 preview 都挂了

**Comment by afc163** [comment]
可能是 token 过期了，新建 PR 是可以的。

File: components/qr-code/index.zh-CN.md [hunk 4]
```diff
@@ -48,8 +49,19 @@ tag: 5.1.0
 | bordered | 是否有边框 | boolean | `true` |
 | errorLevel | 二维码纠错等级 | `'L' \| 'M' \| 'Q' \| 'H' ` | `M` |
 | status | 二维码状态 | `active \| expired \| loading \| scanned` | `active` | scanned: 5.13.0 |
+| statusRender | 自定义状态渲染器 | (oriNode: React.ReactNode, info: [StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode | - | 5.19.0 |
```
Comment (afc163) [inline components/qr-code/index.zh-CN.md:52]:
```suggestion
| statusRender | 自定义状态渲染器 | (originalNode: React.ReactNode, info: [StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode | - | 5.19.0 |
```

File: components/qr-code/index.zh-CN.md [hunk 5]
```diff
@@ -23,6 +23,7 @@ tag: 5.1.0
 <code src="./demo/base.tsx">基本使用</code>
 <code src="./demo/icon.tsx">带 Icon 的例子</code>
 <code src="./demo/status.tsx">不同的状态</code>
+<code src="./demo/customStatusRender.tsx">自定义状态渲染器</code>
```
Comment (li-jia-nan) [inline components/qr-code/index.zh-CN.md:26]:
```suggestion
<code src="./demo/customStatusRender.tsx" version="5.20.0">自定义状态渲染器</code>
```

File: components/qr-code/index.en-US.md [hunk 6]
```diff
@@ -22,6 +22,7 @@ Used when the text needs to be converted into a QR Code.
 <code src="./demo/base.tsx">base</code>
 <code src="./demo/icon.tsx">With Icon</code>
 <code src="./demo/status.tsx">other status</code>
+<code src="./demo/customStatusRender.tsx">custom status render</code>
```
Comment (li-jia-nan) [inline components/qr-code/index.en-US.md:25]:
```suggestion
<code src="./demo/customStatusRender.tsx" version="5.20.0">custom status render</code>
```

File: components/qr-code/hooks/useStatusRender.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    <>
+      <p className={`${prefixCls}-expired`}>{locale?.expired}</p>
+      {onRefresh && (
+        <Button type="link" icon={<ReloadOutlined />} onClick={onRefresh}>
+          {locale?.refresh}
+        </Button>
+      )}
+    </>
+  );
+  const defaultSpin = <Spin />;
```
Comment (li-jia-nan) [inline components/qr-code/hooks/useStatusRender.tsx:32]: 这个是常量，是不是可以放在函数外面？

File: components/qr-code/hooks/useStatusRender.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  };
+
+  const defaultStatusRender: QRCodeProps['statusRender'] = (_oriNode, info) =>
+    defaultNodes[info.status];
+
+  const mergedStatusRender = statusRender ?? defaultStatusRender;
+
+  return {
+    mergedStatusRender,
+    defaultNodes,
```
Comment (afc163) [inline components/qr-code/hooks/useStatusRender.tsx:51]: defaultNodes 貌似没有必要 return 回来。

Comment (MadCcc) [reply components/qr-code/hooks/useStatusRender.tsx:51]: 直接用 mergedStatusRender 执行个结果返回吧

Comment (MadCcc) [reply components/qr-code/hooks/useStatusRender.tsx:51]: 感觉这个 hook 改成 `<QRCodeStatus />` 这样的组件可以

File: components/qr-code/index.en-US.md [hunk 9]
```diff
@@ -47,8 +48,19 @@ Common props ref：[Common props](/docs/react/common-props)
 | bordered | Whether has border style | boolean | `true` |
 | errorLevel | Error Code Level | `'L' \| 'M' \| 'Q' \| 'H' ` | `M` |
 | status | QRCode status | `active \| expired \| loading \| scanned` | `active` | scanned: 5.13.0 |
+| statusRender | custom status render | `(originalNode: React.ReactNode, info: \[StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode` | 5.20.0 |
```
Comment (kiner-tang) [inline components/qr-code/index.en-US.md:51]:
```suggestion
| statusRender | custom status render | `(info: \[StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode` | 5.20.0 |
```

File: components/qr-code/index.zh-CN.md [hunk 10]
```diff
@@ -48,8 +49,19 @@ tag: 5.1.0
 | bordered | 是否有边框 | boolean | `true` |
 | errorLevel | 二维码纠错等级 | `'L' \| 'M' \| 'Q' \| 'H' ` | `M` |
 | status | 二维码状态 | `active \| expired \| loading \| scanned` | `active` | scanned: 5.13.0 |
+| statusRender | 自定义状态渲染器 | (originalNode: React.ReactNode, info: [StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode | - | 5.20.0 |
```
Comment (kiner-tang) [inline components/qr-code/index.zh-CN.md:52]:
```suggestion
| statusRender | 自定义状态渲染器 | (info: [StatusRenderInfo](/components/qr-code-cn#statusrenderinfo)) => React.ReactNode | - | 5.20.0 |
```

**Review by afc163 — APPROVED** [review]

**Pull request merged**